From e6f4de0e55a7ccea74fe24e55d0224558e3a5d0e Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:52:59 -0400
Subject: [fs] ext4: remove i_mutex from ext4_file_sync()

Message-id: <1414058125-4183-26-git-send-email-lczerner@redhat.com>
Patchwork-id: 98133
O-Subject: [RHEL 7.1 PATCH v2 025/171] ext4: remove i_mutex from ext4_file_sync()
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Jan Kara <jack@suse.cz>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit 92e6222dfb85db780ebd8caea6a3f9326c375bc0

 After removal of ext4_flush_unwritten_io() call, ext4_file_sync()
 doesn't need i_mutex anymore. Forcing of transaction commits doesn't
 need i_mutex as there's nothing inode specific in that code apart from
 grabbing transaction ids from the inode. So remove the lock.

 Reviewed-by: Zheng Liu <wenqing.lz@taobao.com>
 Signed-off-by: Jan Kara <jack@suse.cz>
 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/fsync.c b/fs/ext4/fsync.c
index 00e640b..7f7cdda 100644
--- a/fs/ext4/fsync.c
+++ b/fs/ext4/fsync.c
@@ -103,20 +103,19 @@ int ext4_sync_file(struct file *file, loff_t start, loff_t end, int datasync)
   smp_rmb();
   if (EXT4_SB(inode->i_sb)->s_mount_flags & EXT4_MF_FS_ABORTED)
    ret = -EROFS;
-  goto out_trace;
+  goto out;
  }
 
  if (!journal) {
   ret = generic_file_fsync(file, start, end, datasync);
   if (!ret && !hlist_empty(&inode->i_dentry))
    ret = ext4_sync_parent(inode);
-  goto out_trace;
+  goto out;
  }
 
  ret = filemap_write_and_wait_range(inode->i_mapping, start, end);
  if (ret)
   return ret;
- mutex_lock(&inode->i_mutex);
 
  /*
   * data=writeback,ordered:
@@ -148,8 +147,6 @@ int ext4_sync_file(struct file *file, loff_t start, loff_t end, int datasync)
    ret = err;
  }
 out:
- mutex_unlock(&inode->i_mutex);
-out_trace:
  trace_ext4_sync_file_exit(inode, ret);
  return ret;
 }
-- 
1.7.1